File: example/src/main/java/com/github/vivchar/example/pages/simple/DiffUtilFragment.java
package com.github.vivchar.example.pages.simple;

import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.github.vivchar.example.BaseScreenFragment;
import com.github.vivchar.example.R;
import com.github.vivchar.example.widgets.ItemOffsetDecoration;
import com.github.vivchar.rendererrecyclerviewadapter.DefaultDiffCallback;
import com.github.vivchar.rendererrecyclerviewadapter.RendererRecyclerViewAdapter;
import com.github.vivchar.rendererrecyclerviewadapter.ViewModel;
import com.github.vivchar.rendererrecyclerviewadapter.ViewRenderer;

/**
 * Created by Vivchar Vitaly on 28.12.17.
 */

public class DiffUtilFragment extends BaseScreenFragment {

	private YourDataProvider mYourDataProvider = new YourDataProvider();
	private RendererRecyclerViewAdapter mAdapter;
	private RecyclerView mRecyclerView;

	@Nullable
	@Override
	public View onCreateView(final LayoutInflater inflater,
	                         @Nullable final ViewGroup container,
	                         @Nullable final Bundle savedInstanceState) {

		final View view = inflater.inflate(R.layout.fragment_list, container, false);

		mAdapter = new RendererRecyclerViewAdapter();

		mAdapter.setDiffCallback(new DiffCallback());
//		adapter.enableDiffUtil(); /* Or just call it to enable DiffUtil with DefaultDiffCallback */

		mAdapter.registerRenderer(new ViewRenderer<>(R.layout.item_simple_square, DiffViewModel.class,
				(model, finder, payloads) -> finder
						.setText(R.id.text, model.getText())
						.setOnClickListener(R.id.text, v -> {
							reloadItems(model);
						})
		));
//		adapter.registerRenderer(...);
//		adapter.registerRenderer(...);

		mAdapter.setItems(mYourDataProvider.getDiffItems());

		mRecyclerView = (RecyclerView) view.findViewById(R.id.recycler_view);
		mRecyclerView.setAdapter(mAdapter);
		mRecyclerView.setLayoutManager(new GridLayoutManager(getContext(), 4));
		mRecyclerView.addItemDecoration(new ItemOffsetDecoration(10));

		return view;
	}

	private void reloadItems(@NonNull final DiffViewModel model) {
		mAdapter.setItems(mYourDataProvider.getUpdatedDiffItems(model));
	}

	private class DiffCallback extends DefaultDiffCallback<DiffViewModel> {

		@Override
		public boolean areItemsTheSame(@NonNull final DiffViewModel oldItem, @NonNull final DiffViewModel newItem) {
			return oldItem.getID() == newItem.getID();
		}
	}

	public static class DiffViewModel implements ViewModel {

		private final int mID;
		private final String mText;

		public DiffViewModel(final int ID, final String text) {
			mID = ID;
			mText = text;
		}

		public String getText() {
			return mText;
		}

		public int getID() {
			return mID;
		}

		@Override
		public boolean equals(final Object o) {
			if (this == o) {
				return true;
			}
			if (o == null || getClass() != o.getClass()) {
				return false;
			}

			final DiffViewModel that = (DiffViewModel) o;

			if (mID != that.mID) {
				return false;
			}
			return mText != null ? mText.equals(that.mText) : that.mText == null;
		}

		@Override
		public int hashCode() {
			int result = mID;
			result = 31 * result + (mText != null ? mText.hashCode() : 0);
			return result;
		}
	}
}

File: example/src/main/java/com/github/vivchar/example/pages/simple/DiffUtilFragment.kt
package com.github.vivchar.example.pages.simple

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.recyclerview.widget.GridLayoutManager
import com.github.vivchar.example.R
import com.github.vivchar.example.base.BaseFragment
import com.github.vivchar.example.databinding.FragmentListBinding
import com.github.vivchar.example.widgets.ItemOffsetDecoration
import com.github.vivchar.example.widgets.MyAdapter
import com.github.vivchar.rendererrecyclerviewadapter.DefaultDiffCallback
import com.github.vivchar.rendererrecyclerviewadapter.ViewFinder
import com.github.vivchar.rendererrecyclerviewadapter.ViewModel
import com.github.vivchar.rendererrecyclerviewadapter.ViewRenderer
import kotlinx.coroutines.launch

class DiffUtilFragment : BaseFragment<FragmentListBinding>() {

	private val viewModel: DiffUtilViewModel by viewModels()
	private val adapter = MyAdapter()

	override fun createBinding(inflater: LayoutInflater, container: ViewGroup?) =
		FragmentListBinding.inflate(inflater, container, false)

	override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
		super.onViewCreated(view, savedInstanceState)

		adapter.setDiffCallback(DiffCallback())
		adapter.registerRenderer(
			ViewRenderer<DiffViewModel, ViewFinder>(
				R.layout.item_simple_square,
				DiffViewModel::class.java
			) { model, finder, _ ->
				finder
					.setText(R.id.text, model.text)
					.setOnClickListener(R.id.text) { viewModel.onItemClicked(model) }
			}
		)

		binding.recyclerView.adapter = adapter
		binding.recyclerView.layoutManager = GridLayoutManager(context, 4)
		binding.recyclerView.addItemDecoration(ItemOffsetDecoration(10))

		viewLifecycleOwner.lifecycleScope.launch {
			viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
				viewModel.state.collect { state ->
					adapter.setItems(state.items)
				}
			}
		}
	}

	private inner class DiffCallback : DefaultDiffCallback<DiffViewModel>() {
		override fun areItemsTheSame(oldItem: DiffViewModel, newItem: DiffViewModel): Boolean {
			return oldItem.id == newItem.id
		}
	}

	data class DiffViewModel(val id: Int, val text: String) : ViewModel
}
